adafruit_wm8960.py

```diff
@@ -21,48 +25,1587 @@
+    def setLINVOL(self, volume:int) -> bool:
+        # Limit incoming values max
+        if volume > 63:
+            volume = 63
+        if not self._writeRegisterMultiBits(WM8960_REG_LEFT_INPUT_VOLUME, 5, 0, volume): return False
+        return self.pgaLeftIPVUSet()
+
+    '''
+    Sets the volume of the PGA input buffer amp to a specified dB value 
+    passed in as a float argument.
+    Valid dB settings are -17.25 up to +30.00
+    -17.25 = -17.25dB (MIN)
+    ... 0.75dB steps ...
+    30.00 = +30.00dB  (MAX)
+    '''
+    def setLINVOLDB(self, dB:float) -> bool:
+        # Create an unsigned integer volume setting variable we can send to setLINVOL()
+        volume = self.convertDBtoSetting(dB, WM8960_PGA_GAIN_OFFSET, WM8960_PGA_GAIN_STEPSIZE, WM8960_PGA_GAIN_MIN, WM8960_PGA_GAIN_MAX)
+        return self.setLINVOL(volume)
+
+    # 0-63, (0 = -17.25dB) <<-- 0.75dB steps -->> (63 = +30dB)
+    def setRINVOL(self, volume:int) -> bool:
+        # Limit incoming values max
+        if volume > 63:
+            volume = 63
+        if not self._writeRegisterMultiBits(WM8960_REG_RIGHT_INPUT_VOLUME, 5, 0, volume): return False
+        return self.pgaRightIPVUSet()
+
+    '''
+    Sets the volume of the PGA input buffer amp to a specified dB value 
+    passed in as a float argument.
+    Valid dB settings are -17.25 up to +30.00
+    -17.25 = -17.25dB (MIN)
+    ... 0.75dB steps ...
+    30.00 = +30.00dB  (MAX)
+    '''
+    def setRINVOLDB(self, dB:float) -> bool:
+        # Create an unsigned integer volume setting variable we can send to setLINVOL()
+        volume = self.convertDBtoSetting(dB, WM8960_PGA_GAIN_OFFSET, WM8960_PGA_GAIN_STEPSIZE, WM8960_PGA_GAIN_MIN, WM8960_PGA_GAIN_MAX)
+        return self.setRINVOL(volume)
+
+    def setINVOL(self, volume:int) -> bool:
+        return self.setLINVOL(volume) and self.setRINVOL(volume)
+
+    def setINVOLDB(self, dB:float) -> bool:
+        # Create an unsigned integer volume setting variable we can send to setLINVOL()
+        volume = self.convertDBtoSetting(dB, WM8960_PGA_GAIN_OFFSET, WM8960_PGA_GAIN_STEPSIZE, WM8960_PGA_GAIN_MIN, WM8960_PGA_GAIN_MAX)
+        return self.setLINVOL(volume) and self.setRINVOL(volume)
+
+    # Zero Cross prevents zipper sounds on volume changes
+    # Sets both left and right PGAs
+    def enablePgaZeroCross(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 6, 1) and self._writeRegisterBit(WM8960_REG_RIGHT_INPUT_VOLUME, 6, 1)
+    def disablePgaZeroCross(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 6, 0) and self._writeRegisterBit(WM8960_REG_RIGHT_INPUT_VOLUME, 6, 0)
+
+    def enableLINMUTE(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 7, 1)
+    def disableLINMUTE(self) -> bool:
+        return self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 7, 0) and self._writeRegisterBit(WM8960_REG_LEFT_INPUT_VOLUME, 8, 1)
```

Comment:
Instead of have enable and disable functions, please use Adafruit Registers to set bits. https://docs.circuitpython.org/en/latest/docs/design_guide.html#use-adafruit-register-when-possible

Reply:
I'll work on implementing the Register library. It may take some time as I foresee it dramatically changing the layout of this API. For the better, of course!

Reply:
Actually, we may have an issue here. The WM8960 does not support register reads (see page 62 of the datasheet). I don't think that the adafruit_register library will work for this project in its current state.

From page 62: "If the address is not recognised or **the R/W bit is ‘1’**, the WM8960 returns to the idle condition and wait for a new start condition and valid address."

It may be possible to implement an `adafruit_register.i2c_bit.WOBit` class of which you could declare a default value, add a `reset` method, etc, but that would likely cause unexpected results when using multiple bits of the same register on the same device. A per-device buffer of register contents may be able to fix this but might over complicate the adafruit_register library.

I could make a local class which extends RWBit which fixes some of these issues for this specific device, but I am not sure if that is better than the current implementation.

Reply:
The simplest solution may be to use @property and cache the last written value.

Reply:
```python
@property
def left_input_mute(self):
    return self._linmute

@left_input_mute.setter(self, value):
    self._linmute = value
    self._write_register_bit(WM8960_REG_LEFT_INPUT_VOLUME, 7, 1 if value else 0)
```

Reply:
Properties would be an improvement, and I'll go down that road in case adafruit_register doesn't work out.

Here's an potential example of WOBit I whipped up using a static dictionary of memoryviews. Only the first instance of that register will own the original bytearray object.

This won't work if using different devices which utilize this class with overlapping register addresses, but I can't quite figure out how to pull the appropriate register memoryview without requiring that the owner instance be passed in the constructor.

```
class WOBit(RWBit):
    _buffer = {}

    def __init__(
        self,
        register_address: int,
        bit: int,
        register_width: int = 1,
        lsb_first: bool = True,
        default: bool = False,
    ) -> None:
        RWBit.__init__(self, register_address, bit, register_width, lsb_first)
        if not self.buffer[0] in WOBit._buffer:
            WOBit._buffer[register_address] = memoryview(self.buffer)
        else:
            del self.buffer
            self.buffer = WOBit._buffer[register_address]
        self.default = default
        self.reset()
    
    def reset(self) -> None:
        if self.default:
            self.buffer[self.byte] |= self.bit_mask
        else:
            self.buffer[self.byte] &= ~self.bit_mask

    def __get__(
        self,
        obj: Optional[I2CDeviceDriver],
        objtype: Optional[Type[I2CDeviceDriver]] = None,
    ) -> bool:
        return bool(self.buffer[self.byte] & self.bit_mask)

    def __set__(self, obj: I2CDeviceDriver, value: bool) -> None:
        with obj.i2c_device as i2c:
            if value:
                self.buffer[self.byte] |= self.bit_mask
            else:
                self.buffer[self.byte] &= ~self.bit_mask
            i2c.write(self.buffer)
```

Reply:
Further issue here found within the datasheet: "The first 7 bits (B15 to B9) are address bits that select which control register is accessed. The remaining 9 bits (B8 to B0) are data bits, corresponding to the 9 bits in each control register." Using the existing RWBit class will push outside the bounds of the register buffer if the 9th bit is accessed. I will either just convert the existing methods to properties or create a dedicated class for register access based on RWBit for this device.

Reply:
You could probably have a shared buffer on `obj` that all of the descriptors mutate. It is weird but works.

Reply:
That's a great idea. The reset function would likely need to be called within `__init__` and pass `obj` directly, but that can be done.

Reply:
I'm nearly done with my rewrite of the API. There are a large amount of changes and it will likely need to be revised further.

Reply:
Here's the latest commit which addresses this request: https://github.com/adafruit/Adafruit_CircuitPython_WM8960/pull/2/commits/e30892ea3797e32911ad559e1522834004ede7ae

I ended up spinning `RWBit` & `RWBits` off as their own classes, `WMBit` & `WMBits`, as the implementation was far too specific for this device.

I am on the fence with some of the property naming conventions:
- Should "...Enabled" be removed where unnecessary?
- What's the best approach to distinguish between Left & Right properties? For instance, should left/right properties be hidden and stereo properties allow for a tuple to specify channel specific settings?
- For properties which offer a unitless direct or a unit-based assignment (ie: "...Volume" vs "...VolumeDb"), should we instead only offer the unit-based property?
- Should "pga" instead be labeled as "mic" since it is primarily used as a the microphone amplifier?

I've yet to update all of the available examples, but most functionality has been ported over with a few extra properties as I've perused the datasheet.

Reply:
(Just answering questions. Haven't looked at the changes yet.)

> * Should "...Enabled" be removed where unnecessary?

Ya, when it makes sense. The main thing is that properties are named like the thing rather than featuring a verb.

> * What's the best approach to distinguish between Left & Right properties? For instance, should left/right properties be hidden and stereo properties allow for a tuple to specify channel specific settings?

You could just have `left_` and `right_` variables. Tuple properties will require an allocation but can have the advantage that they get set in the same device write.

> * For properties which offer a unitless direct or a unit-based assignment (ie: "...Volume" vs "...VolumeDb"), should we instead only offer the unit-based property?

Try to match units of other devices so that other things could be swapped in. (I'm not sure if there are other existing drivers.)

> * Should "pga" instead be labeled as "mic" since it is primarily used as a the microphone amplifier?

Not sure. I don't know what "pga" means. Words tend to be better than acronyms.

> I've yet to update all of the available examples, but most functionality has been ported over with a few extra properties as I've perused the datasheet.

Reply:
> Ya, when it makes sense. The main thing is that properties are named like the thing rather than featuring a verb.

I'll remove "Enabled" and other verbs where applicable.

> You could just have `left_` and `right_` variables. Tuple properties will require an allocation but can have the advantage that they get set in the same device write.

I'll leave the existing implementation which does this excluding the underscore.

> Try to match units of other devices so that other things could be swapped in. (I'm not sure if there are other existing drivers.)

The only other audio codec library I am aware of is for the VS1053 (https://docs.circuitpython.org/projects/vs1053/en/latest/api.html). I'm not sure if this is the best representation of the policies we want to employ here. Namely, `set_volume(left:int, right:int)` with a range of 0-255. I'd think it would be best to use real-world numbers where possible.

The only downside is when the user would like to control the device on a fully incremental level. Many of the properties of this codec have different decibel minimum and maximums. These would be exposed in API documentation, but the user would have to be aware of the range to utilize the full scope of the property.

> Not sure. I don't know what "pga" means. Words tend to be better than acronyms.

PGA stands for "Programmable-Gain Amplifier". In theory, this input amplifier could be used for a range of devices: electret microphones, high-impedance inputs (ie: a guitar). "mic" designates it as being only usable for 1 thing, which it isn't necessarily the case. Using just "amp" would add confusion with the output mixer and speaker/headphone amps. "inputAmp" may be a viable option, but I'm more inclined to keep it simple with "mic".

Reply:
I believe I've tackled most of the points in this conversation and the API is overall on a much better track. Let me know if you have any further comments.